### PR TITLE
fix: tests failing if user git config not set

### PR DESCRIPTION
**Pull request by listout**
Note for other contributors coming to this issue. I'm getting quite busy with life and it's taking longer than usual to fix this bug. If you have a working patch/fix fell free to open a pull request, I'll close this one.

---

This is WIP fix. The configm_test still needs to be fixed.

### CODE REVIEW

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Hi @listout ,
First of all, thanks a lot for your contribution.

Tbh, I'm not sure which issue you are trying to fix. Could you elaborate a little bit more on the commit message?

**Comment by listout** [comment]
Sorry @rodrigosiqueira for the late reply. I'm trying to fix issue #538. Basically the tests `configm_test` and `explore_test` fail if the global git config (user.name and user.email) are not set.

**Review by kwy95 — CHANGES_REQUESTED** [review]
This first fix seems to work. Please update the commit message as per: https://kworkflow.org/content/howtocontribute.html#commit-messages

**Comment by kwy95** [comment]
@listout I just took a look at the `configm_test` file, and the `tearDown` function removes and recreates the `$SHUNIT_TMPDIR` directory, you can probably circumvent this by also calling `mk_fake_git` after the recreation of the folder.

**Comment by listout** [comment]
Thank you @kwy95. I'll do what you suggested.

**Comment by listout** [comment]
@kwy95 I found a new bug. If you run the explore_test with user config set for signing commits (i.e. gpgsign = true), the test fails.

Steps to reproduce:

- Just set global config for signing commits (` git config --global commit.gpgsign true `)
- Run the explore_test

**Comment by listout** [comment]
Thank you @kwy95. I've changed the commit message. How to get it incorporated into this pull request?

**Comment by listout** [comment]
@kwy95 I tried putting `mk_fake_git` after the `mkdir -p "$SHUNIT_TMPDIR"` command in `tearDown` (line 48 of configm_test.sh) function, same result, test fails.

Log of the test:

```
>> ./run_tests.sh test configm_test
Running test [configm_test]
=========================================================
test_execute_config_manager_SAVE_fails
test_save_config_file_check_save_failures
test_save_config_file_check_directories_creation
test_save_config_file_check_saved_config
test_save_config_file_check_description
test_save_config_file_check_git_save_schema
ASSERT:230: We expected 2 commits, but we got 0
test_save_config_file_check_force
ASSERT:249: We expected no changes
test_list_config_check_when_there_is_no_config
test_list_config_normal_output
test_execute_config_manager_get_config_invalid_option
test_get_config
test_get_config_with_force
test_execute_config_manager_remove_that_should_fail
test_remove_config
ASSERT:432: We expected , 1 files but got 2
test_cleanup
test_get_config_from_proc
test_get_config_from_boot
test_get_config_from_defconfig
test_fetch_config
test_configm_parser

Ran 20 tests.

FAILED (failures=3)
=========================================================
Total: 1 test file(s)
1 test file(s) FAILED

Take a look at:
-> ./tests/configm_test.sh
```

**Comment by rodrigosiqueira** [comment]
Hi @listout ,

I was looking into this `configm_test` and `src/configm`. I noticed that we do not give enough attention to this part of the code, and we are aware of that, as you can see in this issue:

https://github.com/kworkflow/kworkflow/issues/454

Focus on this file and improving it could be an excellent way to get into the project.

Anyway, back to your issue. First of all, `configm_tests` really need some serious review and refactor. I made some changes in how the setUp function works, and I also expanded the `mk_fake_git` to have the gpg configuration. Follows the diff file, but keep in mind that this is just a draft; you probably need to dig into the details.

```diff
diff --git a/tests/configm_test.sh b/tests/configm_test.sh
index 45c54a5..95341c3 100755
diff --git a/tests/configm_test.sh b/tests/configm_test.sh
index 45c54a5..95341c3 100755
--- a/tests/configm_test.sh
+++ b/tests/configm_test.sh
@@ -22,22 +22,31 @@ function setUp()
 {
   local -r current_path="$PWD"
 
+  export KW_CACHE_DIR="$SHUNIT_TMPDIR/cache"
+  mkdir "$SHUNIT_TMPDIR/configs"
+  KW_DATA_DIR="$SHUNIT_TMPDIR"
+  configs_path="$KW_DATA_DIR/configs"
+
   cd "$SHUNIT_TMPDIR" || {
     fail "($LINENO) It was not possible to move to temporary directory"
     return
   }
+
   touch .config
   printf '%s\n' "$CONTENT" > .config
+
+  cd "configs" || {
+    fail "($LINENO) It was not possible to move to configs directory"
+    return
+  }
+
+  mk_fake_git
+
   cd "$current_path" || {
     fail "($LINENO) It was not possible to move back from temp directory"
     return
   }
 
-  export KW_CACHE_DIR="$SHUNIT_TMPDIR/cache"
-  mkdir "$SHUNIT_TMPDIR/configs"
-  KW_DATA_DIR="$SHUNIT_TMPDIR"
-  configs_path="$KW_DATA_DIR/configs"
-
   parse_configuration "$KW_CONFIG_SAMPLE"
   declare -la expected_cmd=()
 }
@@ -214,19 +223,24 @@ function test_save_config_file_check_git_save_schema()
     fail "($LINENO) It was not possible to move to temporary directory"
     return
   }
+
   ret=$(save_config_file $NO_FORCE $NAME_1 "$DESCRIPTION_1")
   ret=$(save_config_file $NO_FORCE $NAME_2 "$DESCRIPTION_2")
+
   cd "configs" || {
     fail "($LINENO) It was not possible to move to configs directory"
     return
   }
+
   ret=$(git rev-list --all --count)
+
   cd "$current_path" || {
     fail "($LINENO) It was not possible to move back from configs directory"
     return
   }
 
-  assertTrue "$LINENO: We expected 2 commits, but we got $ret" '[[ $ret = "2" ]]'
+  # 3 commits from mk_fake_git, 2 commits from this test
+  assertTrue "$LINENO: We expected 5 commits, but we got $ret" '[[ $ret = "5" ]]'
 }
 
 function test_save_config_file_check_force()
diff --git a/tests/utils.sh b/tests/utils.sh
index a40827a..ba1cb6f 100755
--- a/tests/utils.sh
+++ b/tests/utils.sh
@@ -182,6 +182,7 @@ function mk_fake_git()
   git config --local user.name 'Xpto Lala'
   git config --local user.email 'test@email.com'
   git config --local test.config value
+  git config --local commit.gpgsign false
```

**Comment by listout** [comment]
@rodrigosiqueira thank you.

I'm guess it would be alright if I ask questions on here (if I get stuck at some point)?

**Comment by rodrigosiqueira** [comment]
Yeah, feel free to ask questions here or in the issue.

**Comment by rodrigosiqueira** [comment]
Hi @listout ,
Could you rebase your change with the latest code from the unstable branch?

File: README.md [hunk 1]
```diff
@@ -1,13 +1,13 @@
 <img src="images/kw_logo.png" width="600" alt="kworkflow">
 
-![Build Status](https://github.com/kworkflow/kworkflow/actions/workflows/unit_tests.yml/badge.svg?branch=unstable)
-[![codecov](https://codecov.io/gh/kworkflow/kworkflow/branch/unstable/graph/badge.svg?token=ZQ52EK8VWF)](https://codecov.io/gh/kworkflow/kworkflow)
+[![Build Status](https://travis-ci.org/kworkflow/kworkflow.svg?branch=master)](https://travis-ci.org/kworkflow/kworkflow)
```
Comment (kwy95) [inline README.md:3]: Careful when rebasing your changes, these should not be here.

File: tests/explore_test.sh [hunk 2]
```diff
@@ -20,7 +20,7 @@ function setUp()
     return
   }
   # Setup git repository for test
-  git init &> /dev/null
+  mk_fake_git
```
Comment (kwy95) [inline tests/explore_test.sh:23]: This should be the only changed line in this commit.

**Review by kwy95 — CHANGES_REQUESTED** [review]
Be careful when rebasing changes and updating your pull requests, this commit has brought some lines from the `master` branch along with your intended changes.

**Comment by listout** [comment]
Damn, thanks @kwy95. I'll fix it.

**Comment by listout** [comment]
Note for other contributors coming to this issue. I'm getting quite busy with life and it's taking longer than usual to fix this bug. If you have a working patch/fix fell free to open a pull request, I'll close this one.

**Comment by rodrigosiqueira** [comment]
Hi @listout , thanks for your PR and for keeping working in the PR. Could you update it based on @kwy95 comments?

**Comment by listout** [comment]
Hi @rodrigosiqueira, Been quite busy lately, I'll try updating tonight.

**Comment by listout** [comment]
@rodrigosiqueira @kwy95 I've been trying to figure out a way to fix this rebase. All I did was `git rebase` and manually edit the conflicts. 

I'm only familiar with the basics of git any advice would be appreciated.

**Comment by kwy95** [comment]
@listout Hi, sorry for the delayed response, life and work have taken most of my time recently. To fix this PR you can follow these steps:
```bash
 # First fix the commit so it doesn't have those unwanted changes
git reset HEAD~1 # this removes the last commit but the changes are left as unstaged
git add tests/explore_test.sh # add only the relevant file to be committed
git commit # make the desired commit
 # now to remove the unwanted changes you can do
git checkout -- .
 # at this point your branch should be clean and with the desired commit at the tip
 # you can now rebase to update it
git fetch upstream # this gets the latest version of kw
git rebase -i upstream/unstable # this initiates the rebase on top of the unstable branch
 # pick your commit and execute the rebase
 # if there are any conflicts resolve them
 # at this point you should only need to update the github version, so:
git push --force # make sure your branch looks like you want it to as this overwrites the remote
```

**Comment by rodrigosiqueira** [comment]
Hi,
I fixed this issue and merged it with the unstable branch.
Thanks

**Pull request closed without merging**